core/src/main/scala/org/apache/spark/util/ListenerBus.scala

```diff
@@ -17,53 +17,222 @@
 
 package org.apache.spark.util
 
-import java.util.concurrent.CopyOnWriteArrayList
+import java.util.concurrent._
+import java.util.concurrent.atomic.{AtomicBoolean, AtomicLong}
+import java.util.concurrent.locks.{Condition, ReentrantLock}
 
 import scala.collection.JavaConverters._
 import scala.reflect.ClassTag
 import scala.util.control.NonFatal
 
+import com.google.common.util.concurrent.ThreadFactoryBuilder
+
 import org.apache.spark.internal.Logging
+import org.apache.spark.scheduler.LiveListenerBus
+
+private class ListenerEventExecutor[L <: AnyRef] (listenerName: String, queueCapacity: Int)
+  extends Logging {
+  private val threadFactory = new ThreadFactoryBuilder().setDaemon(true)
+    .setNameFormat(listenerName + "-event-executor")
+    .build()
+  val listeners = new CopyOnWriteArrayList[L]()
+  /** Holds the events to be processed by this listener. */
+  private val eventQueue = new LinkedBlockingQueue[Runnable](queueCapacity)
+  /**
+   * A single threaded executor service guarantees ordered processing
+   * of the events per listener.
+   */
+  private val executorService: ThreadPoolExecutor =
+    new ThreadPoolExecutor(1, 1, 0L, TimeUnit.MILLISECONDS, eventQueue, threadFactory)
+  /** A counter for dropped events. It will be reset every time we log it. */
+  private val droppedEventsCounter = new AtomicLong(0L)
+  /** When `droppedEventsCounter` was logged last time in milliseconds. */
+  private var lastReportTimestamp = 0L
+  /** Indicates if we are in the middle of processing some event */
+  private val processingEvent = new AtomicBoolean(false)
+  /**
+   * Indicates if the event executor is started. The executor thread will be
+   * blocked on the condition variable until the event executor is started to
+   * guarantee that we do not process any event before starting the event executor.
+   */
+  private val isStarted = new AtomicBoolean(false)
+  private val lock = new ReentrantLock()
```

Comment:
Minor, but you could achieve the same thing with synchronized/wait/notifyAll, with less code.